### ISSUE TITLE
Warning! Reached max decoder steps

### ISSUE THREAD

**Issue by hadaev8** [NVIDIA/tacotron2#174]
I had no such problem before
But today used pretrained models
And sometimes spectrogram is bugged 

![](http://puu.sh/D77ch/d638c508d8.jpg)

here is notebook for reproduce
https://colab.research.google.com/drive/1jR12cEKdkg0hlDUHGhf2fPb0RwqPwEYj?#scrollTo=CyBu2F7eisFM

**Comment by rafaelvalle** [comment]
Take a look at the gate outputs and decrease the gate threshold accordingly.

**Comment by hadaev8** [comment]
@rafaelvalle 
But I can't get, why spectrogram different for same text?
And why I never notice it before? Last time used in January.
Maybe it because of fc0d34c ?

Also, gate looks this
![](http://puu.sh/D8J3J/7bcb967463.png)

**Comment by rafaelvalle** [comment]
The spectrogram is different with the same text because of the dropout on the Prenet layer: https://github.com/NVIDIA/tacotron2/blob/master/model.py#L100

The gate threshold is what controls when the model should stop decoding. You can decrease the value  to prevent such issues.

**Comment by hadaev8** [comment]
Soo, I get this error with pretrained tacatron and default code from repo, don't you want to investigate?

**Comment by rafaelvalle** [comment]
What sentence did you use?

**Comment by hadaev8** [comment]
Seems it happens sometimes if I have whitespace at the end, like:
'We have started the invasion. '

**Comment by RoelVdP** [comment]
I too think this issue may have been closed out too early. I sometimes see this error for a while, then it is gone again. It **seems** to happen on shorter sentences, or when using just a few words. For example, try "test" as the only input. This happens with pre-trained model. Reducing gate threshold does not help. Increasing decoder steps sometimes works, but it is starting to look somewhat sporadic.

**Comment by RoelVdP** [comment]
I tested/confirmed that removing the change in https://github.com/NVIDIA/tacotron2/commit/fc0d34cfce2c8028a2aa76b535eee73d96cd9a85 makes no difference; the error still happens.

**Comment by RoelVdP** [comment]
Thought; is there any "state" preservation which can be deleted? It seems to be related in some way to a state of the machine, i.e. files present, relevant cache, ...

**Comment by shoegazerstella** [comment]
I am having the same issue. 
My input text is `hello`, [this](https://vocaroo.com/i/s0vvVrgOZfvj) is the -weird- generated wav file. The good thing is that the tts does not crash, but it seems the last letter `o` is repeated for 9 seconds.
Would it be possible to avoid this? maybe by attaching some silence at the end of the phrase if this is too short?

**Comment by rafaelvalle** [comment]
The model is very sensitive to punctuation. Try adding the period: "hello."
You can also try modifying the gate threshold.

**Comment by haqkiemdaim** [comment]
> I am having the same issue.
> My input text is `hello`, [this](https://vocaroo.com/i/s0vvVrgOZfvj) is the -weird- generated wav file. The good thing is that the tts does not crash, but it seems the last letter `o` is repeated for 9 seconds.
> Would it be possible to avoid this? maybe by attaching some silence at the end of the phrase if this is too short?

Hi there @shoegazerstella just want to ask, have you solve the decoder issue ? What would be the best solution ?

**Comment by rafaelvalle** [comment]
Have you tried adding a '.' to the text input: "Hello." instead of "Hello" ?

**Comment by haqkiemdaim** [comment]
> @haqkiemdaim
> Have you tried adding a '.' to the text input: "Hello." instead of "Hello" ?

Haven't yet. But my dataset does not contain any punctuation.

**Comment by rafaelvalle** [comment]
You can add to every sentence of your dataset a token that signifies EOS.

**Comment by haqkiemdaim** [comment]
> You can add to every sentence of your dataset a token that signifies EOS (End of Sentence).

ouhh i see. Is it crucial? because in plain tacotron, i used the same data/corpus without punctuation and it went well. But not in Tacotron2. Any reason for that ?

**Comment by rafaelvalle** [comment]
Tacotron 1 has the EOS token :-)

**Comment by rafaelvalle** [comment]
You can also try adjusting the gate threshold.

**Comment by haqkiemdaim** [comment]
@rafaelvalle This is what i have done so far:
 - remove trailing space 
 - change max decoder steps from 500 to 2000

But still reach max decoder steps at 4th epoch.

Now i'm running again my dataset but with punctuation "." at end of sentence for every line.

Will update here the outcomes.

Regarding the gate threshold, currently it is 0.5.

But im a bit confuse, why we need the gate threshold while we already set the max decoder steps to a certain value ?

**Comment by SomeUserName1** [comment]
@haqkiemdaim Did adding punctuation fix the error?

**Comment by shoegazerstella** [comment]
> @haqkiemdaim Did adding punctuation fix the error?

it worked for my case. thanks!

**Comment by thangdc94** [comment]
@rafaelvalle 
If EOS work why don't you guys make it as default option in `hparams.py`? 
I see other contributors provide answer related to EOS such as accelerate convergence of attention.

**Comment by ErfolgreichCharismatisch** [comment]
How do you try except the message "Warning! Reached max decoder steps"? Which file creates it?
EDIT: I found it at https://github.com/NVIDIA/tacotron2/blob/185cd24e046cc1304b4f8e564734d2498c6e2e6f/model.py

**Comment by coding-dallas** [comment]
> I had no such problem before But today used pretrained models And sometimes spectrogram is bugged
> 
> ![](https://camo.githubusercontent.com/f2c76a11cb343bdb7ee8630c75daefbf384f3a34196b15c8121f5dfcf6065833/687474703a2f2f7075752e73682f44373763682f643633386335303864382e6a7067)
> 
> here is notebook for reproduce https://colab.research.google.com/drive/1jR12cEKdkg0hlDUHGhf2fPb0RwqPwEYj?#scrollTo=CyBu2F7eisFM



> I had no such problem before But today used pretrained models And sometimes spectrogram is bugged
> 
> ![](https://camo.githubusercontent.com/f2c76a11cb343bdb7ee8630c75daefbf384f3a34196b15c8121f5dfcf6065833/687474703a2f2f7075752e73682f44373763682f643633386335303864382e6a7067)
> 
> here is notebook for reproduce https://colab.research.google.com/drive/1jR12cEKdkg0hlDUHGhf2fPb0RwqPwEYj?#scrollTo=CyBu2F7eisFM

the notebook link is not opening
